packages/socket-mode/src/SocketModeClient.ts

```diff
@@ -481,73 +470,164 @@ export class SocketModeClient extends EventEmitter {
+      this.stateMachine.handle(Event.WebSocketOpen, event);
+    });
+    websocket.addEventListener('close', (event) => {
+      this.stateMachine.handle(Event.WebSocketClose, event);
+    });
     websocket.addEventListener('error', (event) => {
-      this.logger.error(`A websocket error occurred: ${event.message}`);
+      this.logger.error(`A WebSocket error occurred: ${event.message}`);
       this.emit('error', websocketErrorWithOriginal(event.error));
     });
     websocket.addEventListener('message', this.onWebSocketMessage.bind(this));
 
-    // Confirm websocket connection is still active
-    websocket.addEventListener('ping', this.heartbeat.bind(this));
+    // Confirm WebSocket connection is still active
+    websocket.addEventListener('ping', ((data: Buffer) => {
+      if (this.pingPongLoggingEnabled) {
+        this.logger.debug(`Received ping from Slack server (data: ${data})`);
+      }
+      this.startMonitoringPingFromSlack();
+      // Since the `addEventListener` method does not accept listener with data arg in TypeScript,
+      // we cast this function to any as a workaround
+    }) as any); // eslint-disable-line @typescript-eslint/no-explicit-any
+
+    websocket.addEventListener('pong', ((data: Buffer) => {
+      if (this.pingPongLoggingEnabled) {
+        this.logger.debug(`Received pong from Slack server (data: ${data})`);
+      }
+      this.lastPongReceivedTimestamp = new Date().getTime();
+      // Since the `addEventListener` method does not accept listener with data arg in TypeScript,
+      // we cast this function to any as a workaround
+    }) as any); // eslint-disable-line @typescript-eslint/no-explicit-any
   }
 
   /**
-   * Tear down method for the client's websocket instance. This method undoes the work in setupWebSocket(url).
+   * Tear down the currently working heartbeat jobs.
    */
-  private teardownWebsocket(): void {
+  private tearDownHeartBeatJobs() {
+    if (this.serverPingTimeout !== undefined) {
+      clearTimeout(this.serverPingTimeout);
+    }
+    if (this.clientPingTimeout !== undefined) {
+      clearTimeout(this.clientPingTimeout);
+    }
+  }
+
+  /**
+   * Tear down method for the client's WebSocket instance.
+   * This method undoes the work in setupWebSocket(url).
+   */
+  private tearDownWebSocket(): void {
     if (this.secondaryWebsocket !== undefined && this.websocket !== undefined) {
-      this.logger.debug('secondary websocket exists, tear down first and assign second');
-      // currently have two websockets, so tear down the older one
+      this.logger.debug('Since the secondary WebSocket exists, going to tear down the first and assign second...');
+      // currently have two WebSocket objects, so tear down the older one
       this.websocket.removeAllListeners('open');
       this.websocket.removeAllListeners('close');
       this.websocket.removeAllListeners('error');
       this.websocket.removeAllListeners('message');
+      this.websocket.close();
```

Comment:
Probably we can improve this switching logic to be a bit safer.

* have a local variable for the old connection and assign the existing connection to it
* switch `this.websocekt` to the new connection
* setTimeout
  * remove the listeners from the old connection
  * close the old one